File: packages/google_mlkit_pose_detection/ios/Classes/GoogleMlKitPoseDetectionPlugin.m
#import "GoogleMlKitPoseDetectionPlugin.h"
#import <MLKitPoseDetection/MLKitPoseDetection.h>
#import <MLKitPoseDetectionCommon/MLKitPoseDetectionCommon.h>
#import <MLKitPoseDetectionAccurate/MLKitPoseDetectionAccurate.h>
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#define channelName @"google_mlkit_pose_detector"
#define startPoseDetector @"vision#startPoseDetector"
#define closePoseDetector @"vision#closePoseDetector"

@implementation GoogleMlKitPoseDetectionPlugin {
    NSMutableDictionary *instances;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitPoseDetectionPlugin* instance = [[GoogleMlKitPoseDetectionPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startPoseDetector]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:closePoseDetector]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (MLKPoseDetector*)initialize:(FlutterMethodCall *)call {
    NSDictionary *options = call.arguments[@"options"];
    NSString *mode = options[@"mode"];
    MLKPoseDetectorMode detectorMode = MLKPoseDetectorModeStream;
    if ([mode isEqualToString:@"single"]) {
        detectorMode = MLKPoseDetectorModeSingleImage;
    }
    
    NSString *model = options[@"model"];
    if ([model isEqualToString:@"base"]) {
        MLKPoseDetectorOptions *options = [[MLKPoseDetectorOptions alloc] init];
        options.detectorMode = detectorMode;
        return [MLKPoseDetector poseDetectorWithOptions:options];
    } else {
        MLKAccuratePoseDetectorOptions *options =
        [[MLKAccuratePoseDetectorOptions alloc] init];
        options.detectorMode = detectorMode;
        return [MLKPoseDetector poseDetectorWithOptions:options];
    }
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    MLKVisionImage *image = [MLKVisionImage visionImageFromData:call.arguments[@"imageData"]];
    
    NSString *uid = call.arguments[@"id"];
    MLKPoseDetector *detector = [instances objectForKey:uid];
    if (detector == NULL) {
        detector = [self initialize:call];
        instances[uid] = detector;
    }
    
    [detector processImage:image
                completion:^(NSArray * _Nullable poses,
                             NSError * _Nullable error) {
        if (error) {
            result(getFlutterError(error));
            return;
        } else if (!poses || poses.count == 0) {
            result(@[]);
            return;
        }
        
        NSMutableArray *array = [NSMutableArray array];
        for (MLKPose *pose in poses) {
            NSMutableArray *landmarks = [NSMutableArray array];
            for (MLKPoseLandmark *landmark in pose.landmarks) {
                NSMutableDictionary *dictionary = [NSMutableDictionary new];
                dictionary[@"type"] = [self poseLandmarkTypeToNumber:landmark.type];
                dictionary[@"x"] = @(landmark.position.x);
                dictionary[@"y"] = @(landmark.position.y);
                dictionary[@"z"] = @(landmark.position.z);
                dictionary[@"likelihood"] = @(landmark.inFrameLikelihood);
                [landmarks addObject:dictionary];
            }
            [array addObject:landmarks];
        }
        result(array);
    }];
}

- (NSNumber*)poseLandmarkTypeToNumber:(MLKPoseLandmarkType)landmarkType {
    NSMutableDictionary *types = [NSMutableDictionary new];
    types[MLKPoseLandmarkTypeNose] = @0;
    types[MLKPoseLandmarkTypeLeftEyeInner] = @1;
    types[MLKPoseLandmarkTypeLeftEye] = @2;
    types[MLKPoseLandmarkTypeLeftEyeOuter] = @3;
    types[MLKPoseLandmarkTypeRightEyeInner] = @4;
    types[MLKPoseLandmarkTypeRightEye] = @5;
    types[MLKPoseLandmarkTypeRightEyeOuter] = @6;
    types[MLKPoseLandmarkTypeLeftEar] = @7;
    types[MLKPoseLandmarkTypeRightEar] = @8;
    types[MLKPoseLandmarkTypeMouthLeft] = @9;
    types[MLKPoseLandmarkTypeMouthRight] = @10;
    types[MLKPoseLandmarkTypeLeftShoulder] = @11;
    types[MLKPoseLandmarkTypeRightShoulder] = @12;
    types[MLKPoseLandmarkTypeLeftElbow] = @13;
    types[MLKPoseLandmarkTypeRightElbow] = @14;
    types[MLKPoseLandmarkTypeLeftWrist] = @15;
    types[MLKPoseLandmarkTypeRightWrist] = @16;
    types[MLKPoseLandmarkTypeLeftPinkyFinger] = @17;
    types[MLKPoseLandmarkTypeRightPinkyFinger] = @18;
    types[MLKPoseLandmarkTypeLeftIndexFinger] = @19;
    types[MLKPoseLandmarkTypeRightIndexFinger] = @20;
    types[MLKPoseLandmarkTypeLeftThumb] = @21;
    types[MLKPoseLandmarkTypeRightThumb] = @22;
    types[MLKPoseLandmarkTypeLeftHip] = @23;
    types[MLKPoseLandmarkTypeRightHip] = @24;
    types[MLKPoseLandmarkTypeLeftKnee] = @25;
    types[MLKPoseLandmarkTypeRightKnee] = @26;
    types[MLKPoseLandmarkTypeLeftAnkle] = @27;
    types[MLKPoseLandmarkTypeRightAnkle] = @28;
    types[MLKPoseLandmarkTypeLeftHeel] = @29;
    types[MLKPoseLandmarkTypeRightHeel] = @30;
    types[MLKPoseLandmarkTypeLeftToe] = @31;
    types[MLKPoseLandmarkTypeRightToe] = @32;
    return types[landmarkType];
}

@end

File: packages/google_mlkit_pose_detection/ios/google_mlkit_pose_detection/Classes/GoogleMlKitPoseDetectionPlugin.swift
import Flutter
import MLKitVision
import MLKitPoseDetection
import MLKitPoseDetectionCommon
import MLKitPoseDetectionAccurate
import google_mlkit_commons

@objc
public class GoogleMlKitPoseDetectionPlugin: NSObject, FlutterPlugin {
  private var instances: [String: PoseDetector] = [:]

  private static let landmarkTypeToNumber: [PoseLandmarkType: Int] = [
    .nose: 0,
    .leftEyeInner: 1,
    .leftEye: 2,
    .leftEyeOuter: 3,
    .rightEyeInner: 4,
    .rightEye: 5,
    .rightEyeOuter: 6,
    .leftEar: 7,
    .rightEar: 8,
    .mouthLeft: 9,
    .mouthRight: 10,
    .leftShoulder: 11,
    .rightShoulder: 12,
    .leftElbow: 13,
    .rightElbow: 14,
    .leftWrist: 15,
    .rightWrist: 16,
    .leftPinkyFinger: 17,
    .rightPinkyFinger: 18,
    .leftIndexFinger: 19,
    .rightIndexFinger: 20,
    .leftThumb: 21,
    .rightThumb: 22,
    .leftHip: 23,
    .rightHip: 24,
    .leftKnee: 25,
    .rightKnee: 26,
    .leftAnkle: 27,
    .rightAnkle: 28,
    .leftHeel: 29,
    .rightHeel: 30,
    .leftToe: 31,
    .rightToe: 32
  ]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "google_mlkit_pose_detector",
      binaryMessenger: registrar.messenger()
    )
    let instance = GoogleMlKitPoseDetectionPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "vision#startPoseDetector":
      handleDetection(call: call, result: result)
    case "vision#closePoseDetector":
      if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
        instances.removeValue(forKey: uid)
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func initialize(call: FlutterMethodCall) -> PoseDetector? {
    guard let args = call.arguments as? [String: Any],
          let optionsDict = args["options"] as? [String: Any] else { return nil }
    let mode = optionsDict["mode"] as? String ?? "stream"
    let detectorMode: PoseDetectorMode = mode == "single" ? .singleImage : .stream
    let model = optionsDict["model"] as? String ?? "base"

    if model == "base" {
      let options = PoseDetectorOptions()
      options.detectorMode = detectorMode
      return PoseDetector.poseDetector(options: options)
    } else {
      let options = AccuratePoseDetectorOptions()
      options.detectorMode = detectorMode
      return PoseDetector.poseDetector(options: options)
    }
  }

  private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let imageData = args["imageData"] as? [String: Any],
          let uid = args["id"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing arguments", details: nil))
      return
    }
    guard let image = VisionImage.visionImage(from: imageData) else {
      result(FlutterError(code: "invalid_image", message: "Invalid or missing image data", details: nil))
      return
    }

    let detector: PoseDetector
    if let existing = instances[uid] {
      detector = existing
    } else {
      guard let newDetector = initialize(call: call) else {
        result(FlutterError(code: "invalid_args", message: "Invalid options", details: nil))
        return
      }
      detector = newDetector
      instances[uid] = detector
    }

    detector.process(image) { poses, error in
      if let error = error as NSError? {
        result(FlutterError(code: "Error \(error.code)", message: error.domain, details: error.localizedDescription))
        return
      }
      guard let poses = poses, !poses.isEmpty else {
        result([])
        return
      }
      let array = poses.map { pose -> [[String: Any]] in
        pose.landmarks.map { landmark in
          [
            "type": Self.landmarkTypeToNumber[landmark.type] ?? -1,
            "x": landmark.position.x,
            "y": landmark.position.y,
            "z": landmark.position.z,
            "likelihood": landmark.inFrameLikelihood
          ] as [String: Any]
        }
      }
      result(array)
    }
  }
}
